### PR TITLE
Update Git mirror support to skip updating the mirror

### PR DESCRIPTION
**Pull request by fatmcgav**
This commit enables the use of an existing Git mirror directory by setting either the `BUILDKITE_GIT_MIRRORS_SKIP_UPDATE` env or the `git-mirrors-skip-update` cli flag to `true`.

## Why?

In a highly dynamic environment, such as using Kubernetes agents, it's not practical to build a Git mirror on every execution. 
However Git mirrors can still be highly beneficial to reduce the overhead of cloning large repositories, so this enables leveraging an existing set of mirror directories mounted via another means, such as a Kubernetes PV or a NFS share, etc.

## Changes

* Added a new `BUILDKITE_GIT_MIRRORS_SKIP_UPDATE` env var and `git-mirrors-skip-update` CLI flag.
* Updated `defaultCheckoutPhase()` logic to leverage new flag.
* Added tests for `dirForRepository()` function.
* Skip using the reference repo if the specified mirror directory doesn't exist.

### CODE REVIEW

**Comment by eleanorakh** [comment]
Thanks so much for opening this up @fatmcgav! This looks interesting; we'll find some time to review it soon. Thank you for your patience and thorough PR description (and tests!). 🙏🏻

**Comment by chloeruka** [comment]
Hey folks, this has been biting us again this week. I hope we can get this looked at soon 🙏🏻

File: bootstrap/bootstrap.go [hunk 1]
```diff
@@ -1279,16 +1279,29 @@ func (b *Bootstrap) defaultCheckoutPhase() error {
...
-			return err
+
+		// Skip updating the Git mirror before using it?
+		if b.Config.GitMirrorsSkipUpdate {
+			mirrorDir = filepath.Join(b.Config.GitMirrorsPath, dirForRepository(b.Repository))
+			// Check if specified mirrorDir exists, otherwise the clone will fail.
+			if !utils.FileExists(mirrorDir) {
+				// Fall back to a clean clone, rather than failing the clone and therefore the build
+				b.shell.Commentf("No existing mirror found for repository %s at %s.", b.Repository, mirrorDir)
+				mirrorDir = ""
```
Comment (jarpy) [inline bootstrap/bootstrap.go:1296]: From my perspective, this is the critical improvement here. We use an external process to provide a shared reference repo to all agents, but we can't guarantee that it's always available. Having the build still run, even if it's in a suboptimal way, is highly desirable!

File: bootstrap/bootstrap.go [hunk 2]
```diff
@@ -1279,16 +1279,29 @@ func (b *Bootstrap) defaultCheckoutPhase() error {
...
 	if experiments.IsEnabled(`git-mirrors`) && b.Config.GitMirrorsPath != "" && b.Config.Repository != "" {
 		b.shell.Commentf("Using git-mirrors experiment 🧪")
-		var err error
-		mirrorDir, err = b.updateGitMirror()
-		if err != nil {
-			return err
+
+		// Skip updating the Git mirror before using it?
+		if b.Config.GitMirrorsSkipUpdate {
+			mirrorDir = filepath.Join(b.Config.GitMirrorsPath, dirForRepository(b.Repository))
```
Comment (moskyb) [inline bootstrap/bootstrap.go:1291]: it'd be nice here to add another comment saying that we're skipping the mirror update

**Review by moskyb — COMMENTED** [review]
one tiny note, but other than that this LGTM! thanks heaps for your contribution :)

File: bootstrap/bootstrap.go (hunk 2 above)
Comment (fatmcgav) [reply bootstrap/bootstrap.go:1292]: Sure, comment added in c36f7946ecbbc8976f3fb13e8707956c717376bf

**Review by moskyb — APPROVED** [review]
LGTM 🎉

I'll merge this and make a release once https://github.com/buildkite/agent/pull/1589 is ready to go :)

**Comment by moskyb** [comment]
hey elastic friends! this will go out in agent [v3.35.0](https://github.com/buildkite/agent/pull/1591), which i'm going to release tomorrow morning new zealand time :)

**Comment by freewil** [comment]
This looks like it may be helpful for my use case, but I'm having a hard time understanding the exact use case for this. Wouldn't you still want to update a git mirror, even if it was mounted via a NFS or similar?

**Comment by fatmcgav** [comment]
> This looks like it may be helpful for my use case, but I'm having a hard time understanding the exact use case for this. Wouldn't you still want to update a git mirror, even if it was mounted via a NFS or similar?

Yeh, updating the mirrors is still required.

In our use case, we have a k8s `Cronjob` that runs every 4 hours that loops through a list of repos to update them...

**Pull request merged**